skopeo: add `--require-signed`

jlebon opened the pull request:
In bootc, we want the ability to assert that signature verification is
enforced.

Add a new top-level `--require-signed` switch. When passed, we use the
new `RequireSignatureVerification()` method to ensure that signature
verification is enforced.

Part of https://github.com/containers/skopeo/issues/1829.

jlebon commented:
This requires https://github.com/containers/container-libs/pull/355 and still needs test.

mtrmac reviewed (commented):
Thanks! Yes, this is the right place to add the option, and overall it looks reasonable.

The option will also need documentation (linters should be complaining — perhaps vendor from your fork, to trigger a CI run).

The immediate next step is to precisely nail down the semantics (and then a matching implementation), that will happen in https://github.com/containers/container-libs/pull/355 .

jlebon committed:
integration/signing_test: move findFingerprint to utils_test.go

Prep for another test which will also use this.

Signed-off-by: Jonathan Lebon <jonathan@jlebon.com>

jlebon committed:
skopeo: add `--require-signed`

In bootc, we want the ability to assert that signature verification is
enforced.

Add a new top-level `--require-signed` switch. When passed, we use the
new `RequireSignatureVerification()` method to ensure that signature
verification is enforced.

Part of https://github.com/containers/skopeo/issues/1829.

Signed-off-by: Jonathan Lebon <jonathan@jlebon.com>

mtrmac commented on integration/require_signed_test.go:
```diff
@@ -0,0 +1,131 @@
+	signaturePath := filepath.Join(srcDir, "signature-1")
+	dockerReference := "localhost/test:latest"
+
+	assertSkopeoSucceeds(t, "", "standalone-sign",
```
(Aesthetically I’m generally not much of a fan of using `standalone-sign`, but here, with the `os.Remove(signaturePath)`, it is a good fit.)

mtrmac reviewed (commented):
Thanks!

Implementation, incl. tests, LGTM — please reuse the setup in `copySuite` to save a bit of test time.

mtrmac commented on integration/require_signed_test.go:
```diff
@@ -0,0 +1,131 @@
+	s.gpgHome = t.TempDir()
+	t.Setenv("GNUPGHOME", s.gpgHome)
+
+	// XXX: probably could have a helper to dedupe this across all the integration tests that use GPG at this point
```
I’m afraid the CI vs. `test-integration` situation is a bit of mess, tracked in https://github.com/containers/skopeo/issues/2701 .

mtrmac reviewed (approved):
Thanks!

cgwalters commented:
This PR came back to the forefront of my mind as I was working on something different related to signing.

So we never did anything to wire this up in bootc side right?

I hate to escalate the scope here...but the thing about a CLI option is while it was easy to implement, it doesn't have a good UX - we don't want users to always need to do `bootc upgrade --require-signed` each time. It needs a config. I think we can probably invent one bootc side, but...

One thing I could imagine here is we try to create the ability to have separate policies for "host" vs "app". Another way I'd like to do this is: "require signatures for privileged containers" - that makes the host not special at all.
